### PR TITLE
[WIP] interfaces/hooks: expose attrs to the interface API, snapctl enhancements (step #4)

### PR DESCRIPTION
**Pull request by stolowski**
Note: https://github.com/snapcore/snapd/pull/3810 has the new API in separate PR and should be reviewed first (it will reduce the size of this diff a bit once landed).

- New ValidatePlug/ValidateSlot methods.
- make attributes persistent (stored in the state) and expose them to the interfaces API.
- support complex structures for attributes via snapctl.
- check attributes set by interface hooks against assertions.

### CODE REVIEW

File: interfaces/policy/policy.go [hunk 1]
```diff
@@ -116,22 +116,23 @@ func (ic *InstallCandidate) Check() error {
...
 	Slot                *snap.SlotInfo
 	SlotSnapDeclaration *asserts.SnapDeclaration
+	SlotAttrs           map[string]interface{}
 
 	BaseDeclaration *asserts.BaseDeclaration
 }
 
 func (connc *ConnectCandidate) plugAttrs() map[string]interface{} {
-	return connc.Plug.Attrs
+	return connc.PlugAttrs
```
Comment (pedronis) [inline interfaces/policy/policy.go]: ConnectCandidate is also meant to write easily tests about the correctness of decls, so I think I would prefer this to have a fallback:

    if connc.PlugAttrs != nil {
        return connc.PlugAttrs
    }
   return connc.Plug.Attrs

and then have in policy just a couple of tests that test the new paths

File: interfaces/policy/policy.go [hunk 2]
```diff
@@ -116,22 +116,23 @@ func (ic *InstallCandidate) Check() error {
...
 }
 
 func (connc *ConnectCandidate) slotAttrs() map[string]interface{} {
-	return connc.Slot.Attrs
+	return connc.SlotAttrs
```
Comment (pedronis) [inline interfaces/policy/policy.go]: same

File: interfaces/policy/policy.go (hunk 1 above)
Comment (pedronis) [inline interfaces/policy/policy.go]: there's just a couple of places where is used for real in ifacestate so the risk of getting it wrong there assuming good tests there should not be big, either way we would need test to check that those fields are set because if they are left empty we get would get wrong behavior either way

Comment (pedronis) [inline interfaces/policy/policy.go]: as I said in the standup with the this change, while we'll need to test the new paths, we can undo the changes to many tests, at least most in policy itself and basedeclaration

File: interfaces/builtin/basedeclaration_test.go [hunk 3]
```diff
@@ -62,6 +62,8 @@ plugs:
 	return &policy.ConnectCandidate{
 		Plug:            plugSnap.Plugs[iface],
 		Slot:            slotSnap.Slots[iface],
+		PlugAttrs:       plugSnap.Plugs[iface].Attrs,
+		SlotAttrs:       slotSnap.Slots[iface].Attrs,
```
Comment (pedronis) [inline interfaces/builtin/basedeclaration_test.go]: this can be undone as well

Comment (stolowski) [inline interfaces/builtin/basedeclaration_test.go]: Done.

**Comment by zyga** [comment]
@stolowski the tests are failing on related code, please have a look

**Comment by stolowski** [comment]
Yes, I'm aware of the test failure; this spread test is abusing content interface and needs some reworking now.

File: interfaces/repo.go [hunk 4]
```diff
@@ -29,6 +29,12 @@ import (
 	"github.com/snapcore/snapd/snap"
 )
 
+// InterfaceAttrs is a container for plug and slot attributes of given connection
+type InterfaceAttrs struct {
```
Comment (zyga) [inline interfaces/repo.go]: Maybe this could be just called `ConnectionAttrs` then?

File: interfaces/repo.go [hunk 5]
```diff
@@ -42,17 +48,21 @@ type Repository struct {
 	// given a plug and a slot, are they connected?
 	plugSlots map[*Plug]map[*Slot]bool
 	backends  map[SecuritySystem]SecurityBackend
+	// attributes of plugs and slots; the attributes include attribute values from
+	// the yaml and provided at runtime via interface hooks.
+	attributes map[*Plug]map[*Slot]*InterfaceAttrs
```
Comment (zyga) [inline interfaces/repo.go]: Just looking at the layout of this. I wonder if we should be following the `slotPlugs` and `plugSlots` so that we can look up information about a connection from either `(plug, slot)` and `(slot, plug)` equally. We could actually store a `connectionInfo` or whatever there so that we could host attributes and know this connection exists in the first place. We already have code managing the two attributes there and we might just re-purpose them for storing connection state.

File: overlord/ifacestate/handlers.go [hunk 6]
```diff
@@ -380,12 +392,45 @@ func (m *InterfaceManager) doConnect(task *state.Task, _ *tomb.Tomb) error {
...
+		if iface == nil {
+			return fmt.Errorf("internal error: cannot find interface: %s", slot.Interface)
+		}
+		type validatePlug interface {
+			ValidatePlug(plug *interfaces.Plug, plugAttrs map[string]interface{}) error
+		}
+		type validateSlot interface {
+			ValidateSlot(slot *interfaces.Slot, slotAttrs map[string]interface{}) error
+		}
+		if validate, ok := iface.(validatePlug); ok {
```
Comment (pedronis) [inline overlord/ifacestate/handlers.go]: it's strange to do this here, it seems repo should grow some ValidateConnection or something like that

File: interfaces/repo.go [hunk 7]
```diff
@@ -438,7 +448,7 @@ func (r *Repository) ResolveDisconnect(plugSnapName, plugName, slotSnapName, slo
 
 // Connect establishes a connection between a plug and a slot.
 // The plug and the slot must have the same interface.
-func (r *Repository) Connect(ref ConnRef) error {
+func (r *Repository) Connect(ref ConnRef, plugAttrs map[string]interface{}, slotAttrs map[string]interface{}) error {
```
Comment (pedronis) [inline interfaces/repo.go]: shouldn't this or somewhere call Sanitize*  on the dynamic attrs, that's the place that can set default value afaiu

also we should remember the issue of not passing nil  to methods that can write on attrs

File: overlord/ifacestate/handlers.go [hunk 8]
```diff
@@ -332,6 +332,18 @@ func (m *InterfaceManager) undoDiscardConns(task *state.Task, _ *tomb.Tomb) erro
 	return nil
 }
 
+func getTaskHookAttributes(task *state.Task) (attrs interfaces.InterfaceAttrs, err error) {
+	attrs.PlugAttrs = make(map[string]interface{})
+	attrs.SlotAttrs = make(map[string]interface{})
+	if err = task.Get("plug-attrs", &attrs.PlugAttrs); err == nil {
```
Comment (pedronis) [inline overlord/ifacestate/handlers.go]: does plug-attrs contains the original attrs from the yaml + what the hook changed? or just new things? the extra  in the arg name to Validate* would make think the latter, but if that's the case the connection check is wrong, that one needs the updated full set or we need to tweak the code in policy a bit differently

File: interfaces/policy/helpers.go [hunk 9]
```diff
@@ -81,15 +81,26 @@ func checkOnClassic(c *asserts.OnClassicConstraint) error {
 }
 
 func checkPlugConnectionConstraints1(connc *ConnectCandidate, cstrs *asserts.PlugConnectionConstraints) error {
```
Comment (pedronis) [inline interfaces/policy/helpers.go]: I don't think we really want to change this code like this, way really need to check the final set of attributes, merged as it makes sense, the rules are about attributes as whole, so checking subsets separetely doesn't have the same meaning

**Comment by chipaca** [comment]
What's the state of this PR?

**Comment by stolowski** [comment]
@chipaca it's WIP after the recent discussions.

**Comment by chipaca** [comment]
removed the 'Decaying' label as you're obviously keeping it up to date. Nevertheless if you're not going to be able to fix those tests soon I'd suggest you close this so we don't get used to seeing it on the queue before it's reviewable again.

**Comment by niemeyer** [comment]
@stolowski I'm closing this as it has plenty of reviews (unclear which ones are done), it says [WIP] in the summary still, it needs conflict solving, and it has a very dirty history:

image

Please open a new one once it's ready for more attention.
